### PR TITLE
4285 timebox create google calendar full sync

### LINKED ISSUE #4285: [Timebox] Create google calendar full-sync
## Scope & Context

We are implementing the google calendar integration.

## Technical inputs

[https://developers.google.com/calendar/api/guides/sync](https://developers.google.com/calendar/api/guides/sync)

Similar to emails.

- Create a `calendar` folder in `packages/twenty-server/src/workspace`.
- Create repositories to interact with calendar data
- Create a google-calendar-full-sync service
- Create a job
- Call the job after the creation of a connected account
- Save **nextSyncToken** after completing full-sync (only present on the last page)

### PR DESCRIPTION
**Pull request by bosiraphael**
Closes #4285

### CODE REVIEW

File: packages/twenty-server/src/workspace/calendar/services/google-calendar-full-sync.service.ts [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+    this.logger.log(
+      `google calendar full-sync for workspace ${workspaceId} and account ${connectedAccountId}: getting existing calendar channel event associations in ${
+        endTime - startTime
+      }ms.`,
+    );
+
+    // TODO: In V2, we will also import deleted events by doing batch GET queries on the canceled events
+    // The canceled events start and end are not accessible in the list query
+    const formattedEvents = events
+      .filter((event) => event.status !== 'cancelled')
```
Comment (Weiko) [inline packages/twenty-server/src/workspace/calendar/services/google-calendar-full-sync.service.ts:146]: Just realised that I think we use "canceled" in the code, right? Is it handled properly?

Comment (Weiko) [reply packages/twenty-server/src/workspace/calendar/services/google-calendar-full-sync.service.ts:146]: It is :) https://github.com/twentyhq/twenty/pull/4442/files#diff-65fa7d566d378d5546e391a1e942301dac777ae21f51cb89f63aec96605c3b52R15

File: packages/twenty-server/src/workspace/calendar-and-messaging/utils/valueStringForBatchRawQuery.util.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+      for (let index = 0; index < numberOfColumns; index++) {
+        value += `$${row * numberOfColumns + index + 1}${
+          typesArray ? `::${typesArray[index]}` : ''
+        }, `;
+      }
+
+      return `(${value.slice(0, -2)})`;
+    })
+    .join(', ');
+};
```
Comment (magrinj) [inline packages/twenty-server/src/workspace/calendar-and-messaging/utils/valueStringForBatchRawQuery.util.ts:21]: I find this more readable that way:

```suggestion
export const valuesStringForBatchRawQuery = (
  values: {
    [key: string]: any;
  }[],
  numberOfColumns: number,
  typesArray: string[] = [],
) => {
  const castedValues = values.reduce((acc, _, rowIndex) => {
    const rowValues = Array.from(
      { length: numberOfColumns },
      (_, columnIndex) => {
        const placeholder = `$${rowIndex * numberOfColumns + columnIndex + 1}`;
        const typeCast = typesArray[columnIndex]
          ? `::${typesArray[columnIndex]}`
          : '';

        return `${placeholder}${typeCast}`;
      },
    ).join(', ');

    acc.push(`(${rowValues})`);

    return acc;
  }, [] as string[]);

  return castedValues.join(', ');
};
```

File: packages/twenty-server/src/workspace/calendar/repositories/calendar-event-attendee/calendar-event-attendee.module.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+
+import { CalendarEventAttendeesService } from 'src/workspace/calendar/repositories/calendar-event-attendee/calendar-event-attendee.service';
+import { WorkspaceDataSourceModule } from 'src/workspace/workspace-datasource/workspace-datasource.module';
+
+@Module({
+  imports: [WorkspaceDataSourceModule],
+  providers: [CalendarEventAttendeesService],
+  exports: [CalendarEventAttendeesService],
+})
+export class CalendarEventAttendeesModule {}
```
Comment (magrinj) [inline packages/twenty-server/src/workspace/calendar/repositories/calendar-event-attendee/calendar-event-attendee.module.ts:11]: Same for the plural naming

File: packages/twenty-server/src/workspace/calendar/repositories/calendar-event/calendar-event.service.ts [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+      calendarEvent.endsAt,
+      calendarEvent.externalCreatedAt,
+      calendarEvent.externalUpdatedAt,
+      calendarEvent.description,
+      calendarEvent.location,
+      calendarEvent.iCalUID,
+      calendarEvent.conferenceSolution,
+      calendarEvent.conferenceUri,
+      calendarEvent.recurringEventExternalId,
+    ]);
```
Comment (magrinj) [inline packages/twenty-server/src/workspace/calendar/repositories/calendar-event/calendar-event.service.ts:142]: If we're flattening all the keys, we can use `Object.keys`

File: packages/twenty-server/src/workspace/calendar/repositories/calendar-event/calendar-event.service.ts [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
```
Comment (magrinj) [inline packages/twenty-server/src/workspace/calendar/repositories/calendar-event/calendar-event.service.ts:194]: Same as above, feel like we can create a util that can create the value string, and create the flat map.
If we need to filter properties we want to flatten we can create a typed param base on the provided object, that allow passing an array of keys.

File: packages/twenty-server/src/workspace/calendar/services/google-calendar-full-sync.service.ts [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+          iCalUIDCalendarEventIdMap,
+          workspaceId,
+          transactionManager,
+        );
+      });
+    } else {
+      this.logger.log(
+        `google calendar full-sync for workspace ${workspaceId} and account ${connectedAccountId} done with nothing to import.`,
+      );
+    }
```
Comment (magrinj) [inline packages/twenty-server/src/workspace/calendar/services/google-calendar-full-sync.service.ts:226]: Feel like if events array is empty we're doing nothing, so I rather put the `if (events.length === 0)` first and then return after the log

File: packages/twenty-server/src/workspace/calendar/services/providers/google-calendar/google-calendar.provider.ts [hunk 7]
```diff
@@ -0,0 +1,42 @@
+import { Injectable } from '@nestjs/common';
+
+import { OAuth2Client } from 'google-auth-library';
+import { calendar_v3, google } from 'googleapis';
```
Comment (magrinj) [inline packages/twenty-server/src/workspace/calendar/services/providers/google-calendar/google-calendar.provider.ts:4]:
```suggestion
import { calendar_v3 as calendarV3, google } from 'googleapis';
```

File: packages/twenty-server/src/workspace/calendar/utils/format-google-calendar-event.util.ts [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+          attendee.responseStatus === 'accepted'
+            ? CalendarEventAttendeeResponseStatus.ACCEPTED
+            : attendee.responseStatus === 'declined'
+              ? CalendarEventAttendeeResponseStatus.DECLINED
+              : attendee.responseStatus === 'tentative'
+                ? CalendarEventAttendeeResponseStatus.TENTATIVE
+                : CalendarEventAttendeeResponseStatus.NEEDS_ACTION,
+      })) || [],
+  };
+};
```
Comment (magrinj) [inline packages/twenty-server/src/workspace/calendar/utils/format-google-calendar-event.util.ts:46]: I think it's more readable that way:

```suggestion
export const formatGoogleCalendarEvent = (
  event: calendar_v3.Schema$Event,
): CalendarEventWithAttendees => {
  const id = v4();

  const formatResponseStatus = (status: CalendarEventAttendeeResponseStatus) => {
    switch (status) {
      case 'accepted':
        return CalendarEventAttendeeResponseStatus.ACCEPTED;
      case 'declined':
        return CalendarEventAttendeeResponseStatus.DECLINED;
      case 'tentative':
        return CalendarEventAttendeeResponseStatus.TENTATIVE;
      default:
        return CalendarEventAttendeeResponseStatus.NEEDS_ACTION;
    }
  };

  return {
    id,
    title: event.summary ?? '',
    isCanceled: event.status === 'cancelled',
    isFullDay: event.start?.dateTime == null,
    startsAt: event.start?.dateTime ?? event.start?.date ?? null,
    endsAt: event.end?.dateTime ?? event.end?.date ?? null,
    externalId: event.id ?? '',
    externalCreatedAt: event.created ?? null,
    externalUpdatedAt: event.updated ?? null,
    description: event.description ?? '',
    location: event.location ?? '',
    iCalUID: event.iCalUID ?? '',
    conferenceSolution:
      event.conferenceData?.conferenceSolution?.key?.type ?? '',
    conferenceUri: event.conferenceData?.entryPoints?.[0]?.uri ?? '',
    recurringEventExternalId: event.recurringEventId ?? '',
    attendees:
      event.attendees?.map((attendee) => ({
        calendarEventId: id,
        iCalUID: event.iCalUID ?? '',
        handle: attendee.email ?? '',
        displayName: attendee.displayName ?? '',
        isOrganizer: attendee.organizer === true,
        responseStatus: formatResponseStatus(attendee.responseStatus),
      })) ?? [],
  };
};
```

File: packages/twenty-server/src/workspace/messaging/services/fetch-by-batch.service.ts [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+import axios, { AxiosInstance, AxiosResponse } from 'axios';
+
+import { BatchQueries } from 'src/workspace/calendar-and-messaging/types/batch-queries';
+import { GmailMessageParsedResponse } from 'src/workspace/messaging/types/gmail-message-parsed-response';
+
+@Injectable()
+export class FetchByBatchesService {
+  private readonly httpService: AxiosInstance;
+
+  constructor() {
```
Comment (magrinj) [inline packages/twenty-server/src/workspace/messaging/services/fetch-by-batch.service.ts:12]: Instead you should inject the httpService in the constructor, and instantiate a new `HttpModule` in the module with the `baseUrl`

**Review by magrinj — COMMENTED** [review]
Some comments :)

File: packages/twenty-server/src/workspace/messaging/repositories/message-channel/message-channel.service.ts [hunk 10]
```diff
@@ -21,7 +21,7 @@ export class MessageChannelService {
       this.workspaceDataSourceService.getSchemaName(workspaceId);
 
     return await this.workspaceDataSourceService.executeRawQuery(
-      `SELECT * FROM ${dataSourceSchema}."messageChannel" WHERE "connectedAccountId" = $1 AND "type" = 'email' LIMIT 1`,
```
[the remark this answers is not in the document]
Comment (bosiraphael) [reply packages/twenty-server/src/workspace/messaging/repositories/message-channel/message-channel.service.ts:24]: It's an error

File: packages/twenty-server/src/workspace/calendar/repositories/calendar-event/calendar-event.service.ts (hunk 4 above)
Comment (bosiraphael) [reply packages/twenty-server/src/workspace/calendar/repositories/calendar-event/calendar-event.service.ts:142]: It has to be in the same order as in the query, also it's not all the keys here

File: packages/twenty-server/src/workspace/calendar/services/google-calendar-full-sync.service.ts (hunk 6 above)
Comment (bosiraphael) [reply packages/twenty-server/src/workspace/calendar/services/google-calendar-full-sync.service.ts:226]: For now yes, but after we will update the sync cursor (commented code below: this.calendarChannelService.updateSyncCursor)

File: packages/twenty-server/src/core/auth/controllers/google-apis-auth.controller.ts [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+
+import { GoogleAPIsProviderEnabledGuard } from 'src/core/auth/guards/google-apis-provider-enabled.guard';
+import { GoogleAPIsOauthGuard } from 'src/core/auth/guards/google-apis-oauth.guard';
+import { GoogleAPIsRequest } from 'src/core/auth/strategies/google-apis.auth.strategy';
+import { GoogleAPIsService } from 'src/core/auth/services/google-apis.service';
+import { TokenService } from 'src/core/auth/services/token.service';
+import { EnvironmentService } from 'src/integrations/environment/environment.service';
+
+@Controller('auth/google-apis')
+export class GoogleAPIsAuthController {
```
[the remark this answers is not in the document]
Comment (bosiraphael) [reply packages/twenty-server/src/core/auth/controllers/google-apis-auth.controller.ts:13]: Yes but in this case I feel like the plural makes more sense, what do you think @Weiko ?

**Review by magrinj — APPROVED** [review]
Stamping it, so we can refactor the folders, we'll post review this PR again then !

**Pull request merged**